### ISSUE TITLE
Filogic: arm-trusted-firmware-tools: compilation broken on Alpine since update to version 2.9

### ISSUE THREAD

**Issue by Shine-** [openwrt/openwrt#13339]
### Describe the bug

Since the upgrade of the arm-trusted-firmware-tools to version 2.9, the compilation for Filogic target fails on Apline Linux. The relevant error message is as follows:

```
  HOSTCC  fiptool.c
fiptool.c: In function 'parse_fip':
fiptool.c:320:39: error: overflow in conversion from 'long unsigned int' to 'int' changes value from '2148012658' to '-2146954638' [-Werror=overflow]
  320 |                 if (ioctl(fileno(fp), BLKGETSIZE64, &st_size) == -1)
      |                                       ^~~~~~~~~~~~
cc1: all warnings being treated as errors
make[4]: *** [Makefile:82: fiptool.o] Error 1
```

I only tested using the 23.05 branch, but I suspect the same happens with Master.
FWIW, I'm using Alpine 3.18.3 in WSL2.
Compilation using other recent distros (e.g. Debian Bookworm) still works fine.

### OpenWrt version

23.05.0-rc3

### OpenWrt target/subtarget

Mediatek / Filogic

### Device

Xiaomi AX6000

### Image kind

Self-built image

### Steps to reproduce

Compile OpenWrt 23.05.0-rc3, 23.05-SNAPSHOT or any revision starting with 6ad85a6c2d733e9b06e0e2b932428cf0d171ff56 for Mediatek Filogic target using the latest Alpine Linux as build host.

### Actual behaviour

Compilation fails with the error message noted above.

### Expected behaviour

Compilation succeeds using any supported distro, including latest Alpine Linux.

### Additional info

n/a

### Diffconfig

```text
n/a
```

### Terms

- [X] I am reporting an issue for OpenWrt, not an unsupported fork.

**Comment by Shine-** [comment]
This works again by now - no idea whether an Alpine update (I'm running 3.19.0 by now) or an OpenWrt commit fixed it.

**Comment by Shine-** [comment]
FWIW, this was apparently never fixed, in fact.

When I closed this, I had probably just accidentally built a device that doesn't use the `arm-trusted-firmware-tools` host package, and therefore I _thought_ that the issue is gone.

In MUSL, the `ioctl` command parameter is declared as `int`, instead of `unsigned`. The `BLKGETSIZE64` `ioctl` command translates to `0x80081272`, ie. the highest bit is set, which is causing the type conflict. Explicitly casting to `int` fixes it here and should not cause any harm for other (glibc) based build hosts either, as the ioctl command is never treated as a number anyway. At least this is my understanding, correct me if I'm wrong.

Keeping this issue closed nevertheless, as it's not an OpenWrt issue.

Adding this comment just in case anybody wants to make OpenWrt build in Alpine again (there are several more build issues present by now, though all rather trivial to solve).

Based on the official Alpine fix for this, here's a patch for OpenWrt:

```diff
diff --git a/package/boot/arm-trusted-firmware-tools/patches/003-fiptool-fix-build-on-musl.patch b/package/boot/arm-trusted-firmware-tools/patches/003-fiptool-fix-build-on-musl.patch
new file mode 100644
index 0000000000..bd50b2410a
--- /dev/null
+++ b/package/boot/arm-trusted-firmware-tools/patches/003-fiptool-fix-build-on-musl.patch
@@ -0,0 +1,11 @@
+--- a/tools/fiptool/fiptool.c
++++ b/tools/fiptool/fiptool.c
+@@ -318,7 +318,7 @@
+ 
+ #ifdef BLKGETSIZE64
+ 	if ((st.st_mode & S_IFBLK) != 0)
+-		if (ioctl(fileno(fp), BLKGETSIZE64, &st_size) == -1)
++		if (ioctl(fileno(fp), (int)BLKGETSIZE64, &st_size) == -1)
+ 			log_err("ioctl %s", filename);
+ #endif
+ 
```

**Comment by Shine-** [comment]
Well, for the record, here are the remaining fixes, as of 25.12-rc4 resp. Main branch from early February 2026.
With the above plus the following, building OpenWrt in Alpine Linux will be possible again.

**bc**: incomplete `getenv()` declaration in `getopt.c`
```diff
diff --git a/tools/bc/patches/001-getopt-declaration.patch b/tools/bc/patches/001-getopt-declaration.patch
new file mode 100644
index 0000000000..157d599a04
--- /dev/null
+++ b/tools/bc/patches/001-getopt-declaration.patch
@@ -0,0 +1,11 @@
+--- a/lib/getopt.c
++++ b/lib/getopt.c
+@@ -197,7 +197,7 @@
+    whose names are inconsistent.  */
+ 
+ #ifndef getenv
+-extern char *getenv ();
++extern char *getenv (const char *);
+ #endif
+ 
+ #endif /* not __GNU_LIBRARY__ */
```

**nftables**: bashism in `configure`
```diff
diff --git a/package/network/utils/nftables/patches/001-for-bash-syntax.patch b/package/network/utils/nftables/patches/001-for-bash-syntax.patch
new file mode 100644
index 0000000000..8db917b5d6
--- /dev/null
+++ b/package/network/utils/nftables/patches/001-for-bash-syntax.patch
@@ -0,0 +1,11 @@
+--- a/configure
++++ b/configure
+@@ -17940,7 +17940,7 @@
+ 	echo "	${STABLE_RELEASE}"
+ 	echo "};"
+ 	echo "static char nftbuildstamp[] = {"
+-	for ((i = 56; i >= 0; i-= 8)); do
++	for i in `seq 56 -8 0`; do
+ 		echo "	((uint64_t)MAKE_STAMP >> $i) & 0xff,"
+ 	done
+ 	echo "};"
```
Note that none of these is an OpenWrt issue.
